### PR TITLE
Binary Search for leftmost or rightmost element

### LINKED ISSUE #10761: Feature Request: Binary Search for leftmost or rightmost element
If I have an array of duplicates, I would like to be able to choose whether or not I want to be able to deterministically search for either the first occurring or last occurring element I am searching for. For example, the below example provides unsatisfactory results.

```chpl
use Search;


var arr1 = [1,2,3,4,5];
var arr2 = [1,2,3,3,3,4,5];
var arr3 = [1,2,3,3,3,3,3,4,5];

writeln(binarySearch(arr1, 3));
writeln(binarySearch(arr2, 3));
writeln(binarySearch(arr3, 3));
```
[TIO](https://tio.run/##S85ILEjN@f@/tDhVITg1sSg5w5qLi6sssUghsajIUMFWIdpQx0jHWMdExzTWGiZuhBA3RpczRpVDyHOVF2WWpObkaSRl5iUWVUIs0wDZoqNgrKlpjVPeiIC8MUT@/38A)

Produces the output...

```
(true, 3)
(true, 4)
(true, 5)
```

I would like to be able to easily specify, perhaps via a parameter, if I want the leftmost, rightmost, or any index that matches the element specified. Something like...

```chpl
enum DuplicateAction {
   FIRST, LAST, ANY
}
proc binarySearch(
   Data: [?Dom], val, comparator: ?rec = defaultComparator, in lo = Dom.low, 
   in hi = Dom.high, param action : DuplicateAction = DuplicateAction.ANY
);
```

### PR DESCRIPTION
**Pull request by dgarvit**
Fixes #10761

### CODE REVIEW

File: modules/packages/Search.chpl [hunk 1]
```diff
@@ -157,7 +167,7 @@ proc linearSearch(Data:[?Dom], val, comparator:?rec=defaultComparator, lo=Dom.al
    :rtype: (`bool`, `Dom.idxType`)
 
  */
-proc binarySearch(Data:[?Dom], val, comparator:?rec=defaultComparator, in lo=Dom.alignedLow, in hi=Dom.alignedHigh) {
+proc binarySearch(Data:[?Dom], val, comparator:?rec=defaultComparator, in lo=Dom.alignedLow, in hi=Dom.alignedHigh, param bound : Bound = Bound.ANY) {
```
Comment (bradcray) [inline modules/packages/Search.chpl]: I've said this several times on the original issue, but I don't see a strong motivation for making this argument a `param` (which will result in copies of the routine getting stamped out for each possible value and prevents a user from making the choice dynamically).  Do you think there's an argument for doing so?  (e.g., will it result in significant increases in execution time that warrants the restricted interface and increase in generated code size?)

**Comment by bradcray** [comment]
Your test will need a .good file in order to work.  You should be able to run `$CHPL_HOME/util/start_test test/.../myTestName.chpl` on the test and get no failures.  See https://github.com/chapel-lang/chapel/blob/master/doc/rst/developer/bestPractices/TestSystem.rst for details.

File: modules/packages/Search.chpl (hunk 1 above)
Comment (dgarvit) [reply modules/packages/Search.chpl]: No, I dont have any particular argument for it. Whats your suggestion? Should I replace `param` with `const` ? Or just let be a variable altogether?

**Comment by dgarvit** [comment]
Yeah, I ran the tests, and got no failures. The `.good` file already exists. The tests are working as expected.

File: modules/packages/Search.chpl (hunk 1 above)
Comment (bradcray) [reply modules/packages/Search.chpl]: I'd leave the argument intent blank (which causes it to be `const in` for enums).

**Comment by bradcray** [comment]
> The .good file already exists.

Oops, my bad.  I was reading too fast, saw there was a .chpl file, but failed to note that you were modifying an existing one rather than adding a new one.  I should go back to my "I won't have time for this this week" statement, as I've clearly proven that to be true.  :)

File: modules/packages/Search.chpl (hunk 1 above)
Comment (dgarvit) [reply modules/packages/Search.chpl]: Made the respective changes :)

**Comment by mppf** [comment]
@bradcray @dgarvit - it's unclear to me what, if anything this PR is waiting for. (e.g. if it's been reviewed and just needs a test run, I could do that...)

**Comment by dgarvit** [comment]
@mppf I think the PR can be merged, it would be great if you could review and test it.

**Comment by bradcray** [comment]
@mppf: The background from my perspective: By itself, the test wasn't enough to convince me that it was working as intended in all cases, and a quick inspection of the code didn't make it clear to me either, so it's been my intention to read the code more carefully and create a more compelling test.  There's also a related design question around what the names and formatting used by the enum should be.  But then more important things seem to always take precedence.  I'm happy to either continue owning this or to have someone else take it over.

**Comment by Rapiz1** [comment]
@bradcray Can I take this and give a more reasonable code snippet?

**Comment by Rapiz1** [comment]
I would like to implement something like `upper_bound` and `lower_bound` in C++.
e.g.
```
enum Bound {lowerBound, upperBound};
proc search(Data:[?Dom], val, comparator:?rec=defaultComparator, lo=Dom.alignedLow, hi=Dom.alignedHigh, sorted=false, bound:Bound) {
  if sorted then
    return binarySearch(Data, val, comparator, lo, hi, bound);
  else
    return linearSearch(Data, val, comparator, lo, hi, bound);
}
```

**Comment by bradcray** [comment]
> Can I take this and give a more reasonable implementation?

I don't know that there's anything unreasonable about @dgarvit's implementation... I just never have the time to give this a review.

I like the notion you suggested of having a general `search()` that did something smart depending on whether the input was already sorted or not, but would probably do that as a follow-on to this PR rather than trying to rewrite this one.  That said, if you check the dates on this, you also may not want to step into the tar pit that it seems to have become in my backlog (suggesting someone else probably should take it over).

**Comment by bradcray** [comment]
Hi @dgarvit — I apologize for sitting on this so long.  The main things that give me pause at this point are (a) the name and capitalization of the enum and its symbols (where I'm trying to think what I think would be better), (b) the lack of documentation for the new argument, (c) the fact that the test case is only a 4-element array, which doesn't seem likely to test corner cases very well.  However, given that this has been in my court for so long, I'm tempted to merge it and then work on it more rather than asking you to do more with it after so long—unless you'd prefer that.

**Comment by bradcray** [comment]
I'm going to close this for the time being due to inactivity (arguably my fault, I'm sorry), with the knowledge that we've got a link to it on the issue requesting it and can come back to it when desire for it heats up again.
